### PR TITLE
ISPN-13445 NearCache factory

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-13445

### CODE REVIEW

**Review by pferraro — APPROVED** [review]
This works for me.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/NearCacheConfigurationBuilder.java [hunk 1]
```diff
@@ -81,6 +84,17 @@ public NearCacheConfigurationBuilder cacheNamePattern(Pattern pattern) {
       return this;
    }
 
+   /**
+    * Specifies a {@link NearCacheFactory} which is reponsible for creating {@link org.infinispan.client.hotrod.near.NearCache} instances.
```
Comment (pruivo) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/NearCacheConfigurationBuilder.java:88]:
```suggestion
    * Specifies a {@link NearCacheFactory} which is responsible for creating {@link org.infinispan.client.hotrod.near.NearCache} instances.
```

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/RemoteCacheConfigurationBuilder.java [hunk 2]
```diff
@@ -96,6 +98,17 @@ public RemoteCacheConfigurationBuilder nearCacheUseBloomFilter(boolean enable) {
       return this;
    }
 
+   /**
+    * Specifies a {@link NearCacheFactory} which is reponsible for creating {@link org.infinispan.client.hotrod.near.NearCache} instances.
```
Comment (pruivo) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/RemoteCacheConfigurationBuilder.java:102]:
```suggestion
    * Specifies a {@link NearCacheFactory} which is responsible for creating {@link org.infinispan.client.hotrod.near.NearCache} instances.
```

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java [hunk 3]
```diff
@@ -232,13 +239,13 @@ public void testNearCacheNamePattern() {
...
-      cacheManager.defineConfiguration("closecache", new org.infinispan.configuration.cache.ConfigurationBuilder().build());
+      cacheManager.defineConfiguration("ncpc", new org.infinispan.configuration.cache.ConfigurationBuilder().build());
       ConfigurationBuilder builder = HotRodClientTestingUtil.newRemoteConfigurationBuilder();
       builder.addServer().host("127.0.0.1").port(hotrodServer.getPort());
-      builder.remoteCache("closecache").nearCacheMode(NearCacheMode.INVALIDATED);
+      builder.remoteCache("ncpc").nearCacheMode(NearCacheMode.INVALIDATED);
       RemoteCacheManager manager = new RemoteCacheManager(builder.build());
       try {
-         RemoteCache nearcache = manager.getCache("closecache");
+         RemoteCache nearcache = manager.getCache("ncpc");
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java:248]: nit
```suggestion
         RemoteCache<?, ?> nearcache = manager.getCache("ncpc");
```

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java [hunk 4]
```diff
@@ -232,13 +239,13 @@ public void testNearCacheNamePattern() {
...
          assertTrue(nearcache instanceof InvalidatedNearRemoteCache);
          RemoteCache cache = manager.getCache();
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java:250]:
```suggestion
         RemoteCache<?, ?> cache = manager.getCache();
```

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java [hunk 5]
```diff
@@ -247,4 +254,87 @@ public void testNearCachePerCache() {
...
 
+   public void testNearCacheFactory() {
+      cacheManager.defineConfiguration("ncf", new org.infinispan.configuration.cache.ConfigurationBuilder().build());
+      ConfigurationBuilder builder = HotRodClientTestingUtil.newRemoteConfigurationBuilder();
+      builder.addServer().host("127.0.0.1").port(hotrodServer.getPort());
+      TestNearCacheFactory testNearCacheFactory = new TestNearCacheFactory();
+      builder.nearCache().mode(NearCacheMode.INVALIDATED).nearCacheFactory(testNearCacheFactory).maxEntries(100);
+      RemoteCacheManager manager = new RemoteCacheManager(builder.build());
+      try {
+         RemoteCache nearcache = manager.getCache("ncf");
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java:265]:
```suggestion
         RemoteCache<String, String> nearcache = manager.getCache("ncf");
```

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java [hunk 6]
```diff
@@ -247,4 +254,87 @@ public void testNearCachePerCache() {
...
+
+   public void testNearCacheFactoryPerCache() {
+      cacheManager.defineConfiguration("ncfpc", new org.infinispan.configuration.cache.ConfigurationBuilder().build());
+      ConfigurationBuilder builder = HotRodClientTestingUtil.newRemoteConfigurationBuilder();
+      builder.addServer().host("127.0.0.1").port(hotrodServer.getPort());
+      TestNearCacheFactory testNearCacheFactory = new TestNearCacheFactory();
+      builder.remoteCache("ncfpc").nearCacheMode(NearCacheMode.INVALIDATED).nearCacheFactory(testNearCacheFactory);
+      RemoteCacheManager manager = new RemoteCacheManager(builder.build());
+      try {
+         RemoteCache nearcache = manager.getCache("ncfpc");
```
Comment (pruivo) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/near/InvalidatedNearCacheTest.java:284]:
```suggestion
         RemoteCache<String, String> nearcache = manager.getCache("ncfpc");
```

**Review by pruivo — COMMENTED** [review]
just small nitpicks here and there :)

**Comment by tristantarrant** [comment]
@pruivo I've applied the fixes

**Review by wburns — APPROVED** [review]

**Review by pruivo — APPROVED** [review]
LGTM. waiting for CI.

**Comment by pruivo** [comment]
failures not related.
Integrated! Thanks @tristantarrant !

**Pull request merged**